src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy

```diff
@@ -331,18 +331,18 @@ public class GenerateProtoTask extends DefaultTask {
   /**
    * Add a directory to protoc's include path.
    */
-  public void addIncludeDir(File dir) {
+  public void addIncludeDir(FileCollection dir) {
     checkCanConfig()
-    includeDirs.add(dir)
+    includeDirs.from(dir)
     // Register all files under the directory as input so that Gradle will check their changes for
     // incremental build
-    inputs.dir(dir).withPathSensitivity(PathSensitivity.RELATIVE)
+    inputs.files(dir).withPathSensitivity(PathSensitivity.RELATIVE)
```

Comment:
What's the difference between `files()` and `dir()`?

Reply:
I believe `inputs.files(includeDirs)` would be the correct one as there could be more than one included directories and any of them changes should trigger the task to run. The documentation is vague and does not mention if calling `inputs.dir()` or `input.file()` multiple times would additively register them as task input or will just replace existing registered one. So `inputs.files(includeDirs)` would be better.

Reply:
`TaskInputs.files()` is unnecessary at least because there is one argument. `TaskInputs.file()`'s doc says it registers one ("some") file, and "The given path is evaluated as per Project.file(Object).". If that "file" is a directory, which is the case here, from our discussion around `FileCollection` (vs `File`) below, it won't be expanded. In contrast, `TaskInputs.dir()` specifically says "All files found under the given directory are treated as input files for this task."

RE: replacing vs addition, all methods in question uses the same verb "Register", which I think implies addition. Even if it did mean "replacing", `files()` and `dir()` should be the same in this aspect.

Reply:
Fair, changed back to `TaskInputs.dir()`.

Reply:
It turns out we cannot use `TaskInputs.file()` or `TaskInputs.dir()` here. As it says, "The given path is evaluated as per Project.file(Object)" for `TaskInputs.file()` and and `TaskInputs.dir()`. This means the actual path (which may not exist) will be resolved as mentioned in [Project.file()](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html#file-java.lang.Object-). If use `TaskInputs.dir()`, we will encounter the same issue as #253 since we are including a non-existing directory in `GenerateProtoTask`'s task inputs.

`.files()` is not only for providing multiple files, but also returns `FileCollection` type, which represents file/directory with a higher abstraction. Since Gradle explicitly resolves the given path for `TaskInputs.file()` and `TaskInputs.dir()`, it provides them as separate APIs for file and directory, and they can only be called on file and directory respectively (otherwise, error). That's the reason, but have nothing to do with whether a directory will be expanded or not. In abstraction, a _directory_ is also a file and anything changing in that directory implies the _directory_ has changed. What "All files found under the given directory are treated as input files for this task" means, I believe, is anything under the given directory is part of task inputs, recursively. Isn't that what we want?

I don't see any problem using registering `FileCollection` as task inputs here, as long as the concern for whether "register" is additive or replacing is clear.

Reply:
I experimented with `TaskInputs.files()` with a single directory path as the input. Gradle is able to detect any changes inside that directory recursively (not just the immediate child of the directory). So I believe using `TaskInputs.files()` for a directory achieves the goal of "Register all files under the directory as input so that Gradle will check their changes for incremental build", which is what we want `TaskInputs.dir()` to do. 
- Seems Gradle has some more advanced ["fingerprint"](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:how_does_it_work) technique to detect task inputs changes, instead of simply based on last-modified-timestamp in Unix. So this can detect changes recursively in a directory tree.

I also verified that `FileCollection` does not expand directory structures if it contains a directory, it's just a set of original files/directories as we give it in.

Also, `TaskInputs.file()`, `TaskInputs.files()` and `TaskInputs.dir()` does register task inputs additively.